### PR TITLE
Revert function pointers

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR reverts #2293, #2316

@DataDog/apm-dotnet

### CODE REVIEW

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs [hunk 1]
```diff
@@ -17,15 +17,12 @@ private static void Inject<TMessageRequest>(SpanContext context, IDictionary mes
             // Consolidate headers into one JSON object with <header_name>:<value>
             var sb = Util.StringBuilderCache.Acquire(Util.StringBuilderCache.MaxBuilderSize);
             sb.Append('{');
-            SpanContextPropagator.Instance.Inject(context, sb, &Setter);
+            SpanContextPropagator.Instance.Inject(context, sb, (StringBuilder carrier, string key, string value) => carrier.AppendFormat("\"{0}\":\"{1}\",", key, value));
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs:20]: I know this is a revert, but does it help to make these static?
```suggestion
            SpanContextPropagator.Instance.Inject(context, sb, static (StringBuilder carrier, string key, string value) => carrier.AppendFormat("\"{0}\":\"{1}\",", key, value));
```

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs [hunk 2]
```diff
@@ -12,12 +12,12 @@ namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.RabbitMQ
 {
     internal static class ContextPropagation
     {
-        public static void HeadersSetter(IDictionary<string, object> carrier, string key, string value)
+        public static readonly Action<IDictionary<string, object>, string, string> HeadersSetter = (carrier, key, value) =>
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs:15]: I'm sure it's _not_ necessary here, but it makes me feel happy

```suggestion
        public static readonly Action<IDictionary<string, object>, string, string> HeadersSetter = static (carrier, key, value) =>
```

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs [hunk 3]
```diff
@@ -12,12 +12,12 @@ namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.RabbitMQ
...
         {
             carrier[key] = Encoding.UTF8.GetBytes(value);
-        }
+        };
 
-        public static IEnumerable<string> HeadersGetter(IDictionary<string, object> carrier, string key)
+        public static readonly Func<IDictionary<string, object>, string, IEnumerable<string>> HeadersGetter = (carrier, key) =>
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs:20]:
```suggestion
        public static readonly Func<IDictionary<string, object>, string, IEnumerable<string>> HeadersGetter = static (carrier, key) =>
```

File: tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingClient.cs [hunk 4]
```diff
@@ -67,7 +67,7 @@ private static void ServiceRemotingClientEvents_SendRequest(object? sender, Even
                         SpanContextPropagator.Instance.Inject(
                             span.Context,
                             messageHeaders,
-                            &HeadersSetter);
+                            (headers, headerName, headerValue) => headers.TryAddHeader(headerName, headerValue));
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingClient.cs:70]:
```suggestion
                            static (headers, headerName, headerValue) => headers.TryAddHeader(headerName, headerValue));
```

File: tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingService.cs [hunk 5]
```diff
@@ -62,7 +62,15 @@ private static void ServiceRemotingServiceEvents_ReceiveRequest(object? sender,
                 // extract propagation context from message headers for distributed tracing
                 if (messageHeaders != null)
                 {
-                    spanContext = SpanContextPropagator.Instance.Extract(messageHeaders, &GetHeaders);
+                    spanContext = SpanContextPropagator.Instance.Extract(messageHeaders, (headers, headerName) =>
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingService.cs:65]:
```suggestion
                    spanContext = SpanContextPropagator.Instance.Extract(messageHeaders, static (headers, headerName) =>
```

File: tracer/src/Datadog.Trace/SpanContextPropagator.cs [hunk 6]
```diff
@@ -27,6 +27,12 @@ internal unsafe class SpanContextPropagator
         private readonly CultureInfo _invariantCulture = CultureInfo.InvariantCulture;
         private readonly IDatadogLogger _log = DatadogLogging.GetLoggerFor<SpanContextPropagator>();
         private readonly ConcurrentDictionary<Key, string?> _defaultTagMappingCache = new();
+        private readonly Func<IReadOnlyDictionary<string, string?>?, string, IEnumerable<string?>> _readOnlyDictionaryValueGetterDelegate;
+
+        private SpanContextPropagator()
+        {
+            _readOnlyDictionaryValueGetterDelegate = (carrier, name) => carrier != null && carrier.TryGetValue(name, out var value) ? new[] { value } : Enumerable.Empty<string?>();
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/SpanContextPropagator.cs:34]:
```suggestion
            _readOnlyDictionaryValueGetterDelegate = static (carrier, name) => carrier != null && carrier.TryGetValue(name, out var value) ? new[] { value } : Enumerable.Empty<string?>();
```

File: tracer/src/Datadog.Trace/SpanContextPropagator.cs [hunk 7]
```diff
@@ -332,5 +329,12 @@ public bool Equals(Key other)
                        TagPrefix == other.TagPrefix;
             }
         }
+
+        private static class DelegateCache<THeaders>
+            where THeaders : IHeadersCollection
+        {
+            public static readonly Func<THeaders, string, IEnumerable<string?>> Getter = (headers, name) => headers.GetValues(name);
+            public static readonly Action<THeaders, string, string?> Setter = (headers, name, value) => headers.Set(name, value);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/SpanContextPropagator.cs:337]:
```suggestion
            public static readonly Func<THeaders, string, IEnumerable<string?>> Getter = static (headers, name) => headers.GetValues(name);
            public static readonly Action<THeaders, string, string?> Setter = static (headers, name, value) => headers.Set(name, value);
```

**Review by andrewlock — APPROVED** [review]

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs (hunk 1 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs (hunk 2 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs (hunk 3 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/RabbitMQ/ContextPropagation.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

File: tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingClient.cs (hunk 4 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingClient.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

File: tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingService.cs (hunk 5 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ServiceFabric/ServiceRemotingService.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

File: tracer/src/Datadog.Trace/SpanContextPropagator.cs (hunk 6 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/SpanContextPropagator.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

File: tracer/src/Datadog.Trace/SpanContextPropagator.cs (hunk 7 above)
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/SpanContextPropagator.cs]: Added in 31aee89. I cannot see any changes in the codegen as expected.

**Review by pierotibou — APPROVED** [review]

**Comment by andrewlock** [comment]
## Benchmarks Report :snail:

Benchmarks for #2331 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c): 
* All benchmarks have **the same** speed
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`WriteAndFlushEnrichedTraces`|net472|860μs|4.06μs|16.7μs|0| 0|0|3.16&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|661μs|2.16μs|8.07μs|0| 0|0|2.57&nbsp;KB|
|#2331|`WriteAndFlushEnrichedTraces`|net472|848μs|1.66μs|5.99μs|0| 0|0|3.16&nbsp;KB|
|#2331|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|670μs|2.69μs|14.2μs|0| 0|0|2.57&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`SendRequest`|net472|0ns|0ns|0ns|0| 0|0|0&nbsp;b|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`SendRequest`|netcoreapp3.1|309μs|1.52μs|6.27μs|0.149| 0|0|19.67&nbsp;KB|
|#2331|`SendRequest`|net472|0ns|0ns|0ns|0| 0|0|0&nbsp;b|
|#2331|`SendRequest`|netcoreapp3.1|299μs|1.38μs|5.71μs|0.151| 0|0|19.67&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`ExecuteNonQuery`|net472|1.66μs|8.83ns|45.9ns|0.0928| 0.000829|0|586&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`ExecuteNonQuery`|netcoreapp3.1|1.47μs|6.38ns|27.1ns|0.00876| 0|0|624&nbsp;B|
|#2331|`ExecuteNonQuery`|net472|1.61μs|7.43ns|28.8ns|0.0929| 0.000815|0|586&nbsp;B|
|#2331|`ExecuteNonQuery`|netcoreapp3.1|1.58μs|6.67ns|24.9ns|0.00879| 0|0|624&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`CallElasticsearch`|net472|2.42μs|12.1ns|58.1ns|0.124| 0|0|794&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`CallElasticsearch`|netcoreapp3.1|1.59μs|7.11ns|27.5ns|0.011| 0|0|784&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`CallElasticsearchAsync`|net472|2.5μs|6.62ns|25.6ns|0.145| 0|0|931&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`CallElasticsearchAsync`|netcoreapp3.1|1.91μs|7.64ns|29.6ns|0.0122| 0|0|904&nbsp;B|
|#2331|`CallElasticsearch`|net472|2.45μs|9.06ns|32.7ns|0.123| 0|0|794&nbsp;B|
|#2331|`CallElasticsearch`|netcoreapp3.1|1.64μs|8.56ns|48.4ns|0.0106| 0|0|784&nbsp;B|
|#2331|`CallElasticsearchAsync`|net472|2.58μs|13ns|55.2ns|0.145| 0|0|931&nbsp;B|
|#2331|`CallElasticsearchAsync`|netcoreapp3.1|1.74μs|7.33ns|31.1ns|0.012| 0|0|904&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`ExecuteAsync`|net472|2.69μs|12.1ns|45.2ns|0.164| 0|0|1.05&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`ExecuteAsync`|netcoreapp3.1|2μs|10.4ns|48.6ns|0.0137| 0|0|1.02&nbsp;KB|
|#2331|`ExecuteAsync`|net472|2.77μs|14.2ns|66.8ns|0.164| 0|0|1.05&nbsp;KB|
|#2331|`ExecuteAsync`|netcoreapp3.1|1.95μs|4.52ns|16.3ns|0.0143| 0|0|1.02&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`SendAsync`|net472|6.13μs|27ns|105ns|0.351| 0|0|2.26&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`SendAsync`|netcoreapp3.1|4.32μs|15.1ns|58.5ns|0.0301| 0|0|2.16&nbsp;KB|
|#2331|`SendAsync`|net472|6.16μs|29.1ns|113ns|0.351| 0|0|2.26&nbsp;KB|
|#2331|`SendAsync`|netcoreapp3.1|4.63μs|14.1ns|52.7ns|0.0301| 0|0|2.16&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|net472|3.39μs|19.3ns|134ns|0.225| 0|0|1.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|netcoreapp3.1|3.22μs|11.2ns|38.6ns|0.0205| 0|0|1.52&nbsp;KB|
|#2331|`EnrichedLog`|net472|3.41μs|12.4ns|48.2ns|0.225| 0|0|1.44&nbsp;KB|
|#2331|`EnrichedLog`|netcoreapp3.1|3.19μs|16.3ns|92ns|0.0202| 0|0|1.52&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|net472|303μs|1.6μs|10.1μs|0.45| 0.15|0|4.31&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|netcoreapp3.1|237μs|803ns|3.11μs|0| 0|0|4.19&nbsp;KB|
|#2331|`EnrichedLog`|net472|296μs|1.03μs|4.01μs|0.437| 0.146|0|4.31&nbsp;KB|
|#2331|`EnrichedLog`|netcoreapp3.1|246μs|1.01μs|3.9μs|0| 0|0|4.19&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|net472|7.72μs|34.7ns|134ns|0.502| 0|0|3.21&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|netcoreapp3.1|6.43μs|26.6ns|99.7ns|0.048| 0|0|3.58&nbsp;KB|
|#2331|`EnrichedLog`|net472|7.92μs|43.1ns|240ns|0.503| 0|0|3.21&nbsp;KB|
|#2331|`EnrichedLog`|netcoreapp3.1|6.59μs|33.3ns|145ns|0.049| 0|0|3.58&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`SendReceive`|net472|2.26μs|11.7ns|53.6ns|0.157| 0|0|1&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`SendReceive`|netcoreapp3.1|2.01μs|6.08ns|21.9ns|0.0143| 0|0|1&nbsp;KB|
|#2331|`SendReceive`|net472|2.2μs|12.4ns|87.6ns|0.158| 0|0|1&nbsp;KB|
|#2331|`SendReceive`|netcoreapp3.1|2μs|6.98ns|27ns|0.013| 0|0|1&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|net472|6.02μs|17.8ns|69ns|0.288| 0|0|1.85&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`EnrichedLog`|netcoreapp3.1|5.15μs|15.7ns|56.7ns|0.0208| 0|0|1.47&nbsp;KB|
|#2331|`EnrichedLog`|net472|6.41μs|32.3ns|141ns|0.288| 0|0|1.85&nbsp;KB|
|#2331|`EnrichedLog`|netcoreapp3.1|5.26μs|17.1ns|66.1ns|0.0209| 0|0|1.47&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`StartFinishSpan`|net472|917ns|4.36ns|16.9ns|0.0704| 0|0|449&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`StartFinishSpan`|netcoreapp3.1|876ns|2.75ns|10.6ns|0.00634| 0|0|448&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`StartFinishScope`|net472|1.13μs|4.72ns|17.7ns|0.0828| 0|0|530&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c)|`StartFinishScope`|netcoreapp3.1|1.07μs|5.78ns|30ns|0.00745| 0|0|568&nbsp;B|
|#2331|`StartFinishSpan`|net472|932ns|4.59ns|18.9ns|0.0706| 0|0|449&nbsp;B|
|#2331|`StartFinishSpan`|netcoreapp3.1|898ns|3.69ns|14.3ns|0.00621| 0|0|448&nbsp;B|
|#2331|`StartFinishScope`|net472|1.13μs|5.44ns|23.1ns|0.0828| 0|0|530&nbsp;B|
|#2331|`StartFinishScope`|netcoreapp3.1|1.04μs|5.67ns|32.1ns|0.00779| 0|0|568&nbsp;B|
</details>

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2331 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c) will **not change** line coverage
:heavy_check_mark: Merging #2331 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c) will **not change** branch coverage
:no_entry: Merging #2331 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c) will will **increase** complexity by `1`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c) | #2331       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `10443` / `14659` | `10404` / `14666` |          |
| Lines %   | `71%`      | `71%`      |  `0%` :heavy_check_mark:  |
| Branches  | `5352` / `7711` | `5330` / `7713` |          |
| Branches %| `69%`      | `69%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `8823`      | `8824`        |  `1`  :no_entry:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=87528&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/87528/artifacts?artifactName=Code%20Coverage%20Report_87528&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2331](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=87625&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/87625/artifacts?artifactName=Code%20Coverage%20Report_87625&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/0875542586063f88ed1da6d3fe875d8b396b7a3c) | #2331       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `71%`      | `71%`       |  `0%` :heavy_check_mark:  |
| Branches %| `69%`      | `69%`       |  `0%` :heavy_check_mark:  |
| Complexity| `8823`      | `8824`       |  `1` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Couchbase.IIOServiceExecuteIntegrationBis](https://github.com/DataDog/dd-trace-dotnet/tree/35e765e82a3bef24b93fa5301b3147c542672a38/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Couchbase/IIOServiceExecuteIntegrationBis.cs) | `-100%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2.TestMethodRunnerExecuteIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/35e765e82a3bef24b93fa5301b3147c542672a38/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodRunnerExecuteIntegration.cs) | `-33%` :no_entry: | `-19%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.NUnit.NUnitCompositeWorkItemSkipChildrenIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/35e765e82a3bef24b93fa5301b3147c542672a38/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/NUnit/NUnitCompositeWorkItemSkipChildrenIntegration.cs) | `-8%` :no_entry: | `-8%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Ci.Agent.CIAgentWriter](https://github.com/DataDog/dd-trace-dotnet/tree/35e765e82a3bef24b93fa5301b3147c542672a38/tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs) | `-5%` :no_entry: | `-21%` :no_entry:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=87528&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/87528/artifacts?artifactName=Code%20Coverage%20Report_87528&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2331](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=87625&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/87625/artifacts?artifactName=Code%20Coverage%20Report_87625&api-version=6.0&%24format=zip)

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs (hunk 1 above)
Comment (lucaspimentel) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs]: > Allow a `static` modifier on lambdas and anonymous methods, which disallows capture of locals or instance state from containing scopes.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/static-anonymous-functions

**Pull request merged**